File: Sources/Gamified/Services/DailyStorageService.swift
//
//  DailyStorageService.swift
//  
//
//  Created by Nathan Fallet on 15/03/2023.
//

import Foundation

public protocol DailyStorageService {
    
    /// Get a value
    /// - Parameters:
    ///   - key: The value to get
    ///   - date: The day to fetch
    /// - Returns: The value for the given day
    func getValue(_ key: String, for date: Date) -> Int
    
    /// Get a values for a set of dates
    /// - Parameters:
    ///   - key: The value to get
    ///   - startDate: The starting date
    ///   - endDate: The ending date
    /// - Returns: The values for the given set of dates
    func getValues(_ key: String, startDate: Date, endDate: Date) -> [Date: Int]
    
    /// Set a value for a day
    /// - Parameters:
    ///   - key: The value to set
    ///   - value: The value to set
    ///   - date: The day to set
    func setValue(_ key: String, value: Int, for date: Date)
    
}

File: Gamified/src/main/java/me/nathanfallet/gamified/services/DailyStorageService.kt
package me.nathanfallet.gamified.services

import java.util.Date

interface DailyStorageService {

    /**
     * Setup the storage to handle a daily value for this key
     * @param key The key for this daily value
     */
    fun setupValue(key: String)

    /**
     * Get a value
     * @param key The value to get
     * @param date The day to fetch
     * @return The value for the given day
     */
    fun getValue(key: String, date: Date): Long

    /**
     * Get a values for a set of dates
     * @param key The value to get
     * @param startDate The starting date
     * @param endDate The ending date
     * @return The values for the given set of dates
     */
    fun getValues(key: String, startDate: Date, endDate: Date): Map<Date, Long>

    /**
     * Set a value for a day
     * @param key The value to set
     * @param value The value to set
     * @param date The day to set
     */
    fun setValue(key: String, value: Long, date: Date)

}